ArmPkg/Drivers/SmmuDxe/SmmuDxe.c

```diff
@@ -0,0 +1,774 @@
+  Cr0.CmdQEn   = 1;
+
+  SmmuV3WriteRegister32 (SmmuInfo->SmmuBase, SMMU_CR0, Cr0.AsUINT32);
+  Status = SmmuV3Poll (SmmuInfo->SmmuBase + SMMU_CR0ACK, 0xC, 0xC);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Error SmmuV3Poll: 0x%lx\n", SmmuInfo->SmmuBase + SMMU_CR0ACK));
+    goto End;
+  }
+
+  //
+  // Invalidate all cached configuration and TLB entries
+  //
+  SMMUV3_BUILD_CMD_CFGI_ALL (&Command);
+  SmmuV3SendCommand (SmmuInfo, &Command);
+  SMMUV3_BUILD_CMD_TLBI_NSNH_ALL (&Command);
+  SmmuV3SendCommand (SmmuInfo, &Command);
+  SMMUV3_BUILD_CMD_TLBI_EL2_ALL (&Command);
+  SmmuV3SendCommand (SmmuInfo, &Command);
+  // Issue a CMD_SYNC command to guarantee that any previously issued TLB
+  // invalidations (CMD_TLBI_*) are completed (SMMUv3.2 spec section 4.6.3).
+  SMMUV3_BUILD_CMD_SYNC_NO_INTERRUPT (&Command);
+  SmmuV3SendCommand (SmmuInfo, &Command);
+
+  // Configure CR0 part2
+  Cr0.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_CR0);
+  ArmDataSynchronizationBarrier ();  // DSB
+
+  Cr0.AsUINT32  = Cr0.AsUINT32 & ~SMMUV3_CR0_VALID_MASK;
+  Cr0.SmmuEn    = 1;
+  Cr0.EventQEn  = 1;
+  Cr0.CmdQEn    = 1;
+  Cr0.PriQEn    = 0;
+  Cr0.Vmw       = 0; // Disable VMID wildcard matching.
+  Idr0.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_IDR0);
+  if (Idr0.Ats != 0) {
+    Cr0.AtsChk = 1;     // disable bypass for ATS translated traffic.
+  }
+
+  SmmuV3WriteRegister32 (SmmuInfo->SmmuBase, SMMU_CR0, Cr0.AsUINT32);
+  Status = SmmuV3Poll (SmmuInfo->SmmuBase + SMMU_CR0ACK, SMMUV3_CR0_SMMU_EN_MASK, SMMUV3_CR0_SMMU_EN_MASK);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Error SmmuV3Poll: 0x%lx\n", SmmuInfo->SmmuBase + SMMU_CR0ACK));
+    goto End;
+  }
+
+  ArmDataSynchronizationBarrier ();  // DSB
+
+  GError.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_GERROR);
+  ASSERT (GError.AsUINT32 == 0);
+
+End:
+  return Status;
+}
+
+/**
+  Retrieve the SMMU configuration data from the HOB.
+
+  @return Pointer to the SMMU_CONFIG structure.
+**/
+STATIC
+SMMU_CONFIG *
+EFIAPI
+GetSmmuConfigHobData (
+  VOID
+  )
+{
+  VOID  *GuidHob;
+
+  GuidHob = GetFirstGuidHob (&gSmmuConfigGuid);
+
+  if (GuidHob != NULL) {
+    return (SMMU_CONFIG *)GET_GUID_HOB_DATA (GuidHob);
+  }
+
+  return NULL;
+}
+
+/**
+  Disable SMMU translation and set Smmu to global bypass during ExitBootServices.
+
+  @param [in] Event    The event that triggered this notification function.
+  @param [in] Context  Pointer to the notification function's context.
+**/
+VOID
+EFIAPI
+SmmuV3ExitBootServices (
+  IN      EFI_EVENT  Event,
+  IN      VOID       *Context
+  )
+{
+  EFI_STATUS  Status;
+
+  Status = SmmuV3DisableTranslation (mSmmu->SmmuBase);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to disable smmu translation.\n", __func__));
+  }
+
+  Status = SmmuV3SetGlobalBypass (mSmmu->SmmuBase);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to set global bypass.\n", __func__));

```

Comment:
Asserts needed for the failures? Could be catastrophic for the OS?

Also a gBS->CloseEvent() should be used here.

Reply:
added asserts and gBS->CloseEVent()